server/core/email.js

```diff
@@ -82,6 +87,62 @@ async function sendVerificationEmail(email, username) {
   return await sendMail(email, subject, html);
 }
 
+async function sendResetPasswordEmail(email, user) {
+  const { username, resetPasswordToken } = user;
+  const token = jwt.sign(
+    { username, resetPasswordToken }, // payload
+    JWT_SECRET, // secret
+    { expiresIn: RESET_PASSWORD_EXPIRATION } // options
+  );
+
+  const subject = 'Reset Password Request';
+  const url = `https://www.watsmymajor.com/reset-password?token=${token}`;
+  const html = `
+  ${emailStyles}
+    <h2>Hi ${username},</h2>
+    <h3>
+      We received a request to reset your WatsMyMajor account password.
+    </h3>
+    <a href="${url}" id="button">Reset Password</a>
+    <p>
+      If you’re having trouble clicking the button, copy and paste the URL below into your web browser.
+    </p>
+    <a href="${url}">${url}</a>
+    <p>
+      If you didn't request a password reset, please ignore this email or <a href="https://github.com/theRoughCode/watsmymajor/issues">let us know</a>.
+    </p>
+  `;
+
+  return await sendMail(email, subject, html);
+}
+
+async function sendResetPasswordSuccess(email, username) {
+  const subject = 'Reset Password Successful';
+  const html = `
+    <h2>Hi ${username},</h2>
+    <h3>
+      Your password was successfully updated!
+    </h3>
+    <p>
+      If you didn't request a password reset, <a href="https://github.com/theRoughCode/watsmymajor/issues">contact us</a>.
+    </p>
+  `;
+
+  return await sendMail(email, subject, html);
+}
+
+// Verify email token to verify user's email
+// Returns { err, username, resetPasswordToken }
+async function verifyResetPasswordToken(token) {
+  try {
+    const { username, resetPasswordToken } = jwt.verify(token, JWT_SECRET);
```

Comment:
do we need an "err" field here?  If so, there could be different types of errors (i.e. expiration).  If it expired, it would be useful to let the user know vs an invalid token.

Reply:
I don't think we need one, since a potential (invalid or expired) token error is handled in the catch block and later in the caller. It's consistent with the other functions (verifyEmailToken and verifyUnwatchToken) too.

Also just realized I didn't include screenshots of error states, but I think this is the only important one. So for now, all we display to the client is:
![image](https://user-images.githubusercontent.com/11607391/57984066-cc364c80-7a25-11e9-9ed6-127337b71eaa.png)

I think not telling the user whether the token expired is valid or expired is sufficient since either way they need to make a new request. (And it may have some security benefits?)

Reply:
Ah I see, thanks for explaining!